### PR TITLE
ARROW-10162: [Rust] Add pretty print support for DictionaryArray

### PR DESCRIPTION
**Pull request by alamb**
This adds some logic for handling dictionary arrays when pretty printing batches. Part of a larger set of work I am working on for better DictionaryArray handling: https://issues.apache.org/jira/browse/ARROW-10159

### CODE REVIEW

File: rust/arrow/src/util/pretty.rs [hunk 1]
```diff
@@ -126,15 +130,56 @@ fn array_value_to_string(column: array::ArrayRef, row: usize) -> Result<String>
...
+            DataType::Int8 => dict_array_value_to_string::<Int8Type>(column, row),
+            DataType::Int16 => dict_array_value_to_string::<Int16Type>(column, row),
+            DataType::Int32 => dict_array_value_to_string::<Int32Type>(column, row),
+            DataType::Int64 => dict_array_value_to_string::<Int64Type>(column, row),
+            DataType::UInt8 => dict_array_value_to_string::<UInt8Type>(column, row),
+            DataType::UInt16 => dict_array_value_to_string::<UInt16Type>(column, row),
+            DataType::UInt32 => dict_array_value_to_string::<UInt32Type>(column, row),
+            DataType::UInt64 => dict_array_value_to_string::<UInt64Type>(column, row),
+            _ => Err(ArrowError::InvalidArgumentError(format!(
+                "Unsupported index type {:?} type for {:?} in repl.",
```
Comment (alamb) [inline rust/arrow/src/util/pretty.rs:143]: I don't really grok why the error messages refer to `repl` in this file, but I have carried forward the tradition with this PR

File: rust/arrow/src/util/pretty.rs [hunk 2]
```diff
@@ -80,8 +84,8 @@ macro_rules! make_string {
     }};
 }
 
-/// Get the value at the given row in an array as a string
-fn array_value_to_string(column: array::ArrayRef, row: usize) -> Result<String> {
+/// Get the value at the given row in an array as a String
+pub fn array_value_to_string(column: &array::ArrayRef, row: usize) -> Result<String> {
```
Comment (alamb) [inline rust/arrow/src/util/pretty.rs:88]: I made this `pub` so I could call it from datafusion/src/test/sql.rs which currently has a copy/pasted version of pretty printing in `array_str`: https://github.com/apache/arrow/blob/master/rust/datafusion/tests/sql.rs#L646

File: rust/arrow/src/util/pretty.rs [hunk 3]
```diff
@@ -60,7 +64,7 @@ fn create_table(results: &[RecordBatch]) -> Result<Table> {
             let mut cells = Vec::new();
             for col in 0..batch.num_columns() {
                 let column = batch.column(col);
-                cells.push(Cell::new(&array_value_to_string(column.clone(), row)?));
```
Comment (alamb) [inline rust/arrow/src/util/pretty.rs:63]: There is no reason / need to clone to column when printing each value

**Comment by jhorstmann** [comment]
Changes look good to me, but I'm wondering now how the existing code handles null values in the primitive arrays. Seems to me that would print the default value. For `Utf8` I think it works kinda implicitly because of how the offsets for null values are layed out.

Would it make sense to fix this in the same PR or should I open a separate ticket?

File: rust/arrow/src/util/pretty.rs (hunk 1 above)
Comment (paddyhoran) [reply rust/arrow/src/util/pretty.rs:143]: I think this was part of DataFusion but was moved into `arrow` as it was generally useful.  I don't understand repl in this context either.  I suggest updating it.

Comment (alamb) [reply rust/arrow/src/util/pretty.rs:143]: I will update the messages.

**Comment by alamb** [comment]
@jhorstmann  

> Changes look good to me, but I'm wondering now how the existing code handles null values in the primitive arrays. Seems to me that would print the default value. For Utf8 I think it works kinda implicitly because of how the offsets for null values are layed out.

Yes you are correct (that the behavior is incorrect). Good point. I opened  https://issues.apache.org/jira/browse/ARROW-10169 and will fix it in a subsequent PR.

**Comment by alamb** [comment]
Rebased to resolve conflicts

File: rust/arrow/src/util/pretty.rs (hunk 2 above)
Comment (alamb) [reply rust/arrow/src/util/pretty.rs:91]: See https://github.com/apache/arrow/pull/8333

**Review by andygrove — APPROVED** [review]

**Pull request closed without merging**